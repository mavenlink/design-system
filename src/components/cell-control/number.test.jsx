import React, { createRef } from 'react';
import { render as _render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Number from './number.jsx';

const render = (ui, options = { labelledBy: 'labelled-by' }) => (
  _render(ui, {
    ...options,
    wrapper: props => (
      <table>
        <thead>
          <tr>
            <th id={options.labelledBy}>Column Header</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            {props.children}
          </tr>
        </tbody>
      </table>
    ),
  })
);

describe('Number cell control', () => {
  const requiredProps = {
    id: 'test-id',
    labelledBy: 'labelled-by',
  };

  it('has defaults', () => {
    const ref = createRef();
    render(<Number {...requiredProps} ref={ref} />);
    expect(document.body).toMatchSnapshot();
    expect(document.body).toBe(document.activeElement);
    expect(ref.current).toMatchSnapshot();
  });

  it('alerts when user focuses into the cell', () => {
    const onFocus = jest.fn(event => event.persist());
    render(<Number onChange={() => {}} value={1} {...requiredProps} onFocus={onFocus} />);

    userEvent.click(within(screen.queryByLabelText('Column Header')).getByDisplayValue('1'));

    expect(onFocus).toHaveBeenCalledWith(expect.objectContaining({ type: 'focus' }));
  });

  it('alerts when user moves focus out of the cell', () => {
    const onBlur = jest.fn(event => event.persist());
    render(<Number onChange={() => {}} value={1} {...requiredProps} onBlur={onBlur} />);

    userEvent.click(within(screen.queryByLabelText('Column Header')).getByDisplayValue('1'));
    userEvent.tab();

    expect(onBlur).toHaveBeenCalledWith(expect.objectContaining({ type: 'blur' }));
  });
});
