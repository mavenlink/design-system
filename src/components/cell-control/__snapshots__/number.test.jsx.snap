// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number cell control has defaults 1`] = `
<body>
  <div>
    <table>
      <thead>
        <tr>
          <th
            id="labelled-by"
          >
            Column Header
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            aria-labelledby="labelled-by"
            class="container"
            role="gridcell"
          >
            <div
              style="height: 100%; position: relative;"
            >
              <input
                aria-describedby="test-id-tooltip test-id-validation-message"
                class="input"
                id="test-id"
                max="2147483648"
                min="-2147483648"
                step="1"
                type="number"
                value=""
              />
              <div
                class="container"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
`;

exports[`Number cell control has defaults 2`] = `
Object {
  "dirty": false,
  "focus": [Function],
  "id": "test-id",
  "name": undefined,
  "validity": ValidityState {},
  "value": undefined,
}
`;
